File: large-class/good/Utility.php
<?php

class Utility {
    public static function validateEmail(string $email): bool {
        return filter_var($email, FILTER_VALIDATE_EMAIL);
    }

    public static function generateToken(): string {
        return bin2hex(random_bytes(32));
    }

    public static function hashPassword(string $password): string {
        return password_hash($password, PASSWORD_DEFAULT);
    }
}

class ActivityLogger {
    private $db;

    public function __construct($db) {
        $this->db = $db;
    }

    public function logActivity(int $userId, string $action): void {
        $stmt = $this->db->prepare("INSERT INTO activity_log (user_id, action, created_at) VALUES (?, ?, NOW())");
        $stmt->execute([$userId, $action]);
    }
}

File: python/large-class/good/utility.py
import hashlib
import secrets

class Utility:
    @staticmethod
    def hash_password(password: str) -> str:
        return hashlib.sha256(password.encode()).hexdigest()

    @staticmethod
    def verify_password(password: str, hashed: str) -> bool:
        return Utility.hash_password(password) == hashed

    @staticmethod
    def validate_email(email: str) -> bool:
        import re
        pattern = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
        return re.match(pattern, email) is not None

    @staticmethod
    def generate_token() -> str:
        return secrets.token_hex(32)


class ActivityLogger:
    def __init__(self, db):
        self.db = db

    def log_activity(self, user_id: int, action: str) -> None:
        cursor = self.db.cursor()
        cursor.execute(
            "INSERT INTO activity_log (user_id, action, created_at) VALUES (?, ?, datetime('now'))",
            (user_id, action)
        )
        self.db.commit()
